chore(deps): Updated to Theia `1.39.0`

kittaakos opened the pull request:
### Motivation
<!-- Why this pull request? -->

Use the latest Theia version, `1.39.0`:

 - bundle backend application with `webpack`,
 - update dependencies to fix security issues,
 - simplify packaging logic,
 - reduce final application size,
 - speed up CI build time (TODO: prove this!)

### Change description
<!-- What does your code do? -->

### Other information
<!-- Any additional information that could help the review process -->

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

kittaakos commented on electron-app/scripts/post-package.js:
```diff
@@ -0,0 +1,186 @@
+    console.log(
+      `🔨  >>> Renaming ${channelFilePath} to ${newChannelFilePath}.`
+    );
+    cpf(channelFilePath, newChannelFilePath);
```
missing `await`

kittaakos replied:
https://github.com/arduino/arduino-ide/pull/2144/commits/501a4229cbe6f5a0870b3586cdbb0e4068313f1d

kittaakos commented on electron-app/scripts/post-package.js:
```diff
@@ -0,0 +1,186 @@
+      await adjustArchiveStructure(fileToCopy, targetFolder);
+    } else {
+      const filename = path.basename(fileToCopy);
+      cpf(fileToCopy, path.join(targetFolder, filename));
```
missing `await`

kittaakos replied:
https://github.com/arduino/arduino-ide/pull/2144/commits/501a4229cbe6f5a0870b3586cdbb0e4068313f1d

kittaakos commented on electron-app/package.json:
```diff
@@ -62,13 +86,137 @@
+    "vscode-builtin-json": "https://open-vsx.org/api/vscode/json/1.46.1/file/vscode.json-1.46.1.vsix",
+    "vscode-builtin-json-language-features": "https://open-vsx.org/api/vscode/json-language-features/1.46.1/file/vscode.json-language-features-1.46.1.vsix",
+    "cortex-debug": "https://downloads.arduino.cc/marus25.cortex-debug/marus25.cortex-debug-1.5.1.vsix",
+    "vscode-language-pack-bg": "https://open-vsx.org/api/MS-CEINTL/vscode-language-pack-bg/1.48.3/file/MS-CEINTL.vscode-language-pack-bg-1.48.3.vsix",
```
update language VSIX

kittaakos replied:
The baseline should be: https://github.com/eclipse-theia/theia/blob/v1.39.0/dev-packages/application-package/src/api.ts#L21

kittaakos replied:
https://github.com/arduino/arduino-ide/commit/661fe1d8cadfa42d81fb71aaee55f6079ba60c5a

kittaakos commented on .vscode/launch.json:
```diff
@@ -95,7 +95,8 @@
       "env": {
-        "TS_NODE_PROJECT": "${workspaceRoot}/tsconfig.json"
+        "TS_NODE_PROJECT": "${workspaceRoot}/tsconfig.json",
+        "IDE2_TEST":"true"
```
https://github.com/arduino/arduino-ide/commit/661fe1d8cadfa42d81fb71aaee55f6079ba60c5a

kittaakos commented:
I have updated the VSIX language packs to VS Code `1.78.0`.  (See [why](https://github.com/arduino/arduino-ide/pull/2144#discussion_r1273245022) `1.78.0`.)

Some screenshots from the IDE2 using Italian as the language. (See some language issues with the `1.37.0` Theia update [here](https://github.com/arduino/arduino-ide/pull/2027#issuecomment-1536398063).)


<img width="404" alt="Screenshot 2023-07-25 at 14 31 32" src="https://github.com/arduino/arduino-ide/assets/1405703/cd52278c-daee-4da1-b07d-d935867af33d">
<img width="553" alt="Screenshot 2023-07-25 at 14 31 28" src="https://github.com/arduino/arduino-ide/assets/1405703/641ee479-4af5-4650-9624-9cba66908049">
<img width="399" alt="Screenshot 2023-07-25 at 14 31 22" src="https://github.com/arduino/arduino-ide/assets/1405703/31d564be-bf12-41e4-acc9-02dd6834a72c">
<img width="303" alt="Screenshot 2023-07-25 at 14 31 08" src="https://github.com/arduino/arduino-ide/assets/1405703/60cee9d6-1c08-4a07-bcca-3ec783b92d8c">
<img width="309" alt="Screenshot 2023-07-25 at 14 31 01" src="https://github.com/arduino/arduino-ide/assets/1405703/bc6ae4f2-2722-44f7-8f4b-1b5b34432092">

Please review. Thank you!

kittaakos commented:
Must be aligned with https://github.com/arduino/arduino-ide/pull/2151.

AlbyIanna reviewed (approved):
I tested as many feature as I could on MacOs (Ventura 13.4) and couldn't spot any issue! LGTM ✅

per1234 reviewed (changes requested):
Incorrect layout of editor tab dialog error message highlight

Selecting "**New Tab**" or "**Rename**" (if a secondary sketch file tab is selected) from the editor toolbar context menu causes a dialog to open. The user input in the tab name field is validated. If the validation fails, an error message is printed in the dialog.

🐛 The position of the colored background on this text is off-center and doesn't have any padding for the dialog's buttons:

![image](https://github.com/arduino/arduino-ide/assets/8572152/a6729d9a-19e5-46f5-813f-a69c1aaa5d65)

### To reproduce

1. Select "**New Tab**" from the editor toolbar context menu.
1. In the "**Name for new file**" field, type `foo bar`

The expected error message appears:

> The name must start with a letter, number, or underscore, followed by letters, numbers, dashes, dots and underscores. Maximum length is 63 characters.

🐛 but the position of the colored background on this text is off-center and doesn't have any padding for the dialog's buttons.

### Expected behavior

The error message highlight is centered on the text and has padding against the buttons.

**-OR-**

The error message is not highlighted.

### Arduino IDE version

090098d (tester build for e8a9ef7f49b215b8c862bad45ad9f518ff7ca92b)

### Operating system

Windows 11

### Additional context

The highlighting behavior is something introduced by the PR. There is no such highlight on the error message when using the build from the `main` branch (f6a4325):

![image](https://github.com/arduino/arduino-ide/assets/8572152/d8fa5787-2972-4770-807d-e6ffcd42e959)

I think it is useful to highlight the error message in some way (though I would prefer it was done in a consistent manner in all the dialogs e.g.,  Preferences dialog when invalid sketchbook location is selected, Firmware Updater dialog when process fails) to bring it to the attention of the user. However, if fixing the layout is more difficult, I am fine with the alternative fix of removing the highlight altogether as that would not be a regression.

per1234 reviewed (changes requested):
**UPDATE**: Fixed by https://github.com/arduino/arduino-ide/commit/129436bc8a77336fdd0c14444bbd3486b04f4579

<s>

## Deleting to trash fails

The file of the currently selected editor tab can be deleted by selecting "**Delete**" from the editor toolbar context menu. The IDE defaults to deleting this via the operating system's "trash" system in order to allow the user to recover the file.

🐛 The attempt to delete to "trash" fails and the user is forced to use the fallback of permanent deletion.

### To reproduce

1. Select "**New Tab**" from the editor toolbar context menu.
   The "**Name for new file**" dialog will open.
1. Type `Foo` in the field of the "**Name for new file**" dialog.
1. Click the "**OK**" button.
1. Select "**Delete**" from the editor toolbar context menu.
   The "**Move File to Trash**" dialog will open.
1. Click the "**OK**" button in the "**Move File to Trash**" dialog.

🐛 The attempt to delete the file fails:

> Error deleting file
>
>Failed to delete "Foo.ino" using the Trash. Do you want to permanently delete instead?
>
>You can disable the use of Trash in the preferences.

### Expected behavior

Deletion via the "trash" system works.

### Arduino IDE version

090098d (tester build for e8a9ef7f49b215b8c862bad45ad9f518ff7ca92b)

### Operating system

Windows 11

### Additional context

The following is printed to the logs when the deletion fails:

```text
2023-07-30T12:08:32.365Z root ERROR Error deleting with trash: Error: Request 'delete' failed
    at Proxy.<anonymous> (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:3032271)
    at b.delete (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:3741400)
    at x.doDelete (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:3664211)
    at async x.delete (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:3663346)
    at async d.moveFileToTrash (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:8661047)
    at async Promise.all (index 1)
    at async d.delete (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:8660793)
    at async Promise.all (index 0)
    at async d.execute (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:8659279)
    at async g.executeCommand (file:///E:/incoming/review/2144/1-090098d/resources/app/lib/frontend/bundle.js:2:2954564)
Caused by: EntryNotFound (FileSystemError): Error: spawn E:\incoming\review\2144\1-090098d\resources\app\lib\backend\windows-trash.exe ENOENT
    at u (E:\incoming\review\2144\1-090098d\resources\app\lib\backend\main.js:2:1306074)
    at T.toFileSystemProviderError (E:\incoming\review\2144\1-090098d\resources\app\lib\backend\main.js:2:1329976)
    at T.delete (E:\incoming\review\2144\1-090098d\resources\app\lib\backend\main.js:2:1324688)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async g.onRequest (E:\incoming\review\2144\1-090098d\resources\app\lib\backend\main.js:2:1064552)
    at async c.handleRequest (E:\incoming\review\2144\1-090098d\resources\app\lib\backend\main.js:2:1055445)
```

---

I used a secondary sketch file in the demo. If the primary sketch file is selected, the deletion fails silently.

---

Deletion to trash works as expected when using the IDE build from the `main` branch (f6a4325).

</s>

kittaakos commented:
> I think it is useful to highlight the error message in some way

Thanks for spotting it! I agree, and I will take care of it in another PR. For now, I reverted (https://github.com/arduino/arduino-ide/pull/2144/commits/3529b147f5f05d650560338f826762e7f374da5a) the changes IDE2 pulled in from upstream:


<img width="554" alt="Screenshot 2023-07-30 at 17 45 15" src="https://github.com/arduino/arduino-ide/assets/1405703/9dec45ad-1148-43e2-8d79-1fe8646e88d0">

kittaakos commented:
> Deleting to trash fails

Thanks for the excellent review. See https://github.com/eclipse-theia/theia/issues/12780. I am looking for a quick solution.

per1234 reviewed (changes requested):
**UPDATE**: Resolved by https://github.com/arduino/arduino-ide/commit/e506b6f46be5616ccf7a2f15ca3c3a2cee467abe

<s>

## "User Data Directory" folder name changed from `arduino-ide` to `electron-app`

Arduino IDE stores some data files in a special folder referred to as the "User Data Directory". The folder has historically been at this path:

```text
%APPDATA%\arduino-ide\
```
(i.e., `C:\Users\<username>\AppData\Roaming\arduino-ide\`)


The folder is now at this path:

```text
%APPDATA%\electron-app\
```

🐛 The data that was stored in the previous location will no longer be used by Arduino IDE.
🐛 Advanced users who work directly with this folder (e.g., to simulate the first run behavior for beta testing) will be confused by the change in behavior.
🐛 The generic `electron-app` name doesn't identify the folder as being associated with the Arduino IDE application.

### To reproduce

1. Select **File > Quit** from the Arduino IDE menus if it is running.
1. Delete the folders at the following paths:
   ```text
   C:\Users\<username>\AppData\Roaming\arduino-ide\
   ```
   ```text
   C:\Users\<username>\AppData\Roaming\electron-app\
   ```
1. Start Arduino IDE.
1. Check the contents of the `C:\Users\<username>\AppData\Roaming` folder.

🐛 A folder named `electron-app` was created instead of the expected `arduino-ide` name.

### Expected behavior

"User data directory" name is `arduino-ide`

### Arduino IDE version

40c251c (tester build for dc8360bb6c6d146f20cae505669b33d64d05ee63)

### Operating system

Windows 11

### Additional context

The build from the `main` branch (e17472e) uses the `arduino-ide` name.

---

I guess the source of the name is here:

https://github.com/arduino/arduino-ide/blob/dc8360bb6c6d146f20cae505669b33d64d05ee63/electron-app/package.json#L3

instead of here as before:

https://github.com/arduino/arduino-ide/blob/dc8360bb6c6d146f20cae505669b33d64d05ee63/package.json#L2

---

When I check the command line of the Arduino IDE process, I see this path is passed via the `--user-data-dir` flag.

</s>

kittaakos commented:
IDE2 will lose the default value (`'nightly'`) generated for the `arduino.ide.updateChannel` preference. See https://github.com/arduino/arduino-ide/issues/2157.

kittaakos commented:
> "User data directory" name is `arduino-ide`

This should work now with the latest build from this PR. Resolved via https://github.com/arduino/arduino-ide/pull/2144/commits/e506b6f46be5616ccf7a2f15ca3c3a2cee467abe. I have verified it on macOS; the path was slightly different: `/Users/<username>/Library/Application Support`. Please double-check it on Windows. Thank you!

per1234 commented on arduino-ide-extension/package.json:
```diff
@@ -170,17 +171,17 @@
   ],
   "arduino": {
-    "cli": {
+    "arduino-cli": {
```
This line must also be updated to reflect the key name change:

https://github.com/arduino/arduino-ide/blob/28bc470f6b13b87f1c60606c06b7d65a3c82bd5f/arduino-ide-extension/src/node/config-service-impl.ts#L135

Otherwise the "CLI Version" field in the "About" dialog is empty:

![image](https://github.com/arduino/arduino-ide/assets/8572152/022f6f5a-081b-4d51-b6d4-d11069d0ef48)

per1234 replied:
Fixed by https://github.com/arduino/arduino-ide/commit/75480bd53f247fb214075339a6a55cf757adaab5

Thanks!

per1234 reviewed (changes requested):
**UPDATE**: Fixed by https://github.com/arduino/arduino-ide/commit/75480bd53f247fb214075339a6a55cf757adaab5

<s>

## "Date" field of "about" dialog uses fallback

The build date is shown in the "Date" field of the "about" dialog. The dialog code is written so that it falls back on using the value of "dev build" in this field if the build date data is not available.

🐛 This fallback value is being used in the field in the standard beta tester build generated by the repository's continuous development system.

### To reproduce

1. Download the [beta tester build for the PR](https://github.com/arduino/arduino-ide/blob/main/docs/contributor-guide/beta-testing.md#testing-pull-requests).
1. Start the beta tester build.
1. Select **Help > About Arduino IDE** from Arduino IDE menus.
   The "**Arduino IDE**" dialog will open.

🐛 The "**Date**" field is set to "dev build":

![image](https://github.com/arduino/arduino-ide/assets/8572152/c576fc03-3233-4ad4-93c6-c3dd2ba16aa6)

### Expected behavior

The "**Date**" field is always set to the build timestamp when the build is generated by standard processes.

### Arduino IDE version

3fc51f4 (tester build for 28bc470f6b13b87f1c60606c06b7d65a3c82bd5f)

### Operating system

Windows 11

### Additional context

I haven't found the data from this field to be essential, so if there is some difficulty in populating it specifically for beta tester builds then I am OK with it being set to "dev build" for these builds. My concern is that this is an indicator that the system for populating the field is completely broken and it will be set to this fallback value even for the release builds, which would make the field useless and also possibly cause the users confusion.

</s>

kittaakos commented:
I do not know what's been wrong with GitHub recently, but the commits do not appear on the PR.

Here is my commit (https://github.com/arduino/arduino-ide/commit/75480bd53f247fb214075339a6a55cf757adaab5) on the top of the branch, but it's missing from the PR, and the build has not run either.


<img width="734" alt="Screenshot 2023-08-01 at 20 38 30" src="https://github.com/arduino/arduino-ide/assets/1405703/b9e65403-8646-4075-b84d-c4f0de71d718">

<img width="1288" alt="Screenshot 2023-08-01 at 20 41 11" src="https://github.com/arduino/arduino-ide/assets/1405703/5f5030a4-5b06-4da9-8e5c-f70427c744ee">


Anyway, if the build starts, please check the fixes. I verified it with a locally bundled app on macOS, and it worked. Thanks!

per1234 reviewed (changes requested):
**UPDATE**: Fixed by https://github.com/arduino/arduino-ide/pull/2144/commits/920a4fc16fb78beb0c3647426480a6e45ba90f1e

<s>

When I perform the instructions under "Build from source" in `docs/development.md` using a fresh clone of the repository checked out to this PR branch (f7cb8aabdc04dd52e4ca787dff04dba48abdc2d0), step (4) fails:

```text
Failed to start the electron application.
Error: Cannot find module './cli-protocol/cc/arduino/cli/settings/v1/settings_pb'
Require stack:
- /mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/node/config-service-impl.js
- /mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/node/sketches-service-impl.js
- /mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/electron-main/theia/electron-main-application.js
- /mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/electron-main/electron-arduino.js
- /mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/electron-main/arduino-electron-main-module.js
- /mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/electron-app/src-gen/backend/electron-main.js
- /mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/node_modules/electron/dist/resources/default_app.asar/main.js
- 
    at Module._resolveFilename (node:internal/modules/cjs/loader:1002:15)
    at n._resolveFilename (node:electron/js2c/browser_init:2:109734)
    at Module._load (node:internal/modules/cjs/loader:848:27)
    at f._load (node:electron/js2c/asar_bundle:2:13330)
    at Module.require (node:internal/modules/cjs/loader:1068:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/node/config-service-impl.js:50:23)
    at Module._compile (node:internal/modules/cjs/loader:1174:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1229:10)
    at Module.load (node:internal/modules/cjs/loader:1044:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/node/config-service-impl.js',
    '/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/node/sketches-service-impl.js',
    '/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/electron-main/theia/electron-main-application.js',
    '/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/electron-main/electron-arduino.js',
    '/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/arduino-ide-extension/lib/electron-main/arduino-electron-main-module.js',
    '/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/electron-app/src-gen/backend/electron-main.js',
    '/mnt/1c0bd221-e077-452c-b688-160dc700c122/git/arduino-ide_fresh-clone/node_modules/electron/dist/resources/default_app.asar/main.js',
    undefined
  ]
}
```

There is no `arduino-ide-extension/lib/node/cli-protocol` folder.

I don't have this problem when I run the same procedure in the repository where I had previously ran Arduino IDE from source with the repository checked out to the `main` branch.

</s>

kittaakos commented on docs/development.md:
````diff
@@ -57,25 +57,75 @@ Please refer to the [Theia IDE prerequisites](https://github.com/eclipse-theia/t
-    ```sh
-    yarn
-    ```
+If you want to develop the application, do the followings:
````
https://github.com/arduino/arduino-ide/pull/2144/commits/7ead2d9f5ceb506258a9204d81ef278557237aa9

kittaakos commented on docs/development.md:
````diff
@@ -57,25 +57,75 @@ Please refer to the [Theia IDE prerequisites](https://github.com/eclipse-theia/t
+3. Package the application
+   ```sh
+   yarn --cwd electron-app package
+   ```
````
Thanks! Yes, it makes sense: https://github.com/arduino/arduino-ide/pull/2144/commits/920a4fc16fb78beb0c3647426480a6e45ba90f1e

kittaakos commented on docs/development.md:
```diff
@@ -57,25 +57,75 @@ Please refer to the [Theia IDE prerequisites](https://github.com/eclipse-theia/t
+
+2. Rebuild the native dependencies for electron
+
+   - <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd> to open the _Command Palette_.
```
https://github.com/arduino/arduino-ide/pull/2144/commits/920a4fc16fb78beb0c3647426480a6e45ba90f1e

kittaakos commented:
Thank you for the review! I have addressed the outstanding change requests and, most importantly, fixed the dev docs by adding the missing step. It should work now.

per1234 reviewed (approved):
Thanks Akos!
